README.md
# COVID-19 SIMULATION

This is an agent-based covid-19 simulation that can be used to discover which strategies can contain the covid-19 outbreak and minimize damage to our freedoms.

Although the simulation is based on a toy model, the positive effects of certain strategies should be applicable to more realistic models.

Discussions of some experiments run with this simulation can be found at https://www.forastelo.org/covid-19. New experiments and articles will be added in the future.

# Notes
- The model is very simple, ignores many details and is still being improved.
- The code is work in progress, needs lots of refactoring and optimization. Some portions of the code are not thoroughly tested and should not be used. See 'How to use the simulation' section.

# Requirements
The code has been run successfully with the following configuration:
- python >= 3.6.5 (64-bit)
- matplotlib 3.2.1
- numpy 1.18.2

# How to use the simulation
## Running a single simulation
- Navigate to the root of the repository
- Change the parameters of the sample simulation in [run_single.py](run_single.py) file
- Run the script
```bash
python run_single.py
```

Simulation will show how the state of the society changes in time and will create a summary graph with important metrics.

## Running an experiment and analyzing the results
- Modify the sample experiment in the [experiments.py](experiments.py) file
    - The resulting graph is optimized for 2 simulations, so it is better to compare only 2 simulations in a single experiment
    - As the world size and number of simulations get larger, it takes much more time to run an experiment
- Navigate to the root of the repository and run the experiment script
```bash
python run_experiment.py
```

Once the simulations end, a json output file will be generated in the results folder.

- Execute 'run_analysis.py' script to see an analysis of the results
```bash
python run_analysis.py
```
The results will be displayed as a bar chart.
This script analyzes the latest file by default. In order to analyze other files, you need to change the script and pass a list of file names as the argument.

# How to configure a simulation
- More information will be here soon.
- Read the comments in the __init__method of the Simulation class in [simulation.py](covid19_sim/simulation.py) file to learn more about the parameters.
- There will be more sample scenarios in the future.
- Do NOT use TestingStrategy.SCAN and TestingStrategy.SELF_CHECK as they are not yet well tested.

# Assumptions, rules and definitions
- Each day, each person interacts with up to 8 nearby contacts unless they are in quarantine. Each person's interactions happen one after another but each person is selected randomly from the population.
- There is a probability for an infected person to infect another person in each interaction.
- If air travel is enabled, locations of some people are swapped with each other to simulate air travel. The probability of getting infected during air travel is related to the percentage of infected people in the population.
- By default, 20% of all cases are assumed to be asymptomatic.
- Deaths and fatality of the disease are ignored. It is assumed that everything else being equal, minimizing the number of infected people and maximum infection rate will minimize deaths.
- Infected people recover from the disease between 14 and 21 days. Both symptomatic and asymptomatic cases become recovered/immune when they recover.
- When a symptomatically infected person recovers, the person is 'identified as immune' by the society. However, when an asymptomatic case recovers, they are NOT 'identified as immune' because there is no way to know whether or not they had the disease.
- Right now, when a person is tested positive but has no symptoms (asymptomatic case), they are NEVER identified as immune by the society even if they are removed from quarantine.
- Contact tracing
    - There is a probability for a symptomatically infected person to go to hospital. There is also a probability for uninfected people to go to hospital with similar symptoms.
    - There is a probability to trace the contacts of a person. If the probability is lower than 100%, some of the contacts will not be traced at all.
    - There is a probability to reach the traced contacts of a person on a given day. A traced contact may or may not be available for testing on a given day.
    - There is a probability for a traced contact to be put in quarantine. When a contact is traced, it may or may not be possible (or desirable) to put the contact in quarantine until testing takes place.
    - If an infected person does long distance travel and goes to hospital, the person's contacts before the travel are not traced.
    - When a person goes to hospital, they are put in a sick queue. Their contacts are put in a contacts queue.
    - People in the sick queue are tested first. If the queue is empty, people in the contacts queue are tested. If a contact is positive, their contacts are traced.
    - Everyday, there is a maximum testing capacity. If the number of people in the queues are larger than the testing capacity, some people may not be tested on that day.
    - If scanning option is on, all the people who are not in any test queues and who are not identified as immune are put in a screening queue. Initially, people are ordered starting with the top left corner of the map, going downwards row by row.
    - The screening queue is an ordered list. People who are not tested and not in quarantine have higher priority. Previously tested people are also put in the queue after a certain time that can be specified as a parameter. Those whose last test dates are further in the past are given priority.
    - People in the screening queue are tested only if there is excess testing capacity on a given day.
- People who are infected 1 day ago are counted as new infections. Obviously, this information is not visible to the society, but the assumption is that no matter how society counts new infections, that number will be proportional to the real number of new infections.
- Unnecessarily quarantined people
    - When an uninfected person is quarantined, this is counted as an 'unnecessary quarantine'. Although it can be argued that isolating uninfected people is not unnecessary because it prevents them from catching the disease and later infecting others, this term makes sense from an ideal viewpoint: if the society could perfectly detect the infected people, they would be put in quarantine without the uninfected people having to give up their freedoms. So, it is important to keep in mind that 'unnecessary quarantine' is defined from that viewpoint. This metric measures the ineffectiveness of the society to detect the infected people and chase the virus. Therefore, the smaller is this metric, the better is the strategy in finding and isolating infected people.
- Unnecessary quarantine days
    - This is the total number of days lost to unnecessary quarantine. The larger is this metric, the more uninfected people lose their freedoms because of quarantines.
- By experimenting with different parameters and rules, it is possible to optimize the outcome and reduce infection rates and days lost to unnecessary quarantine.

# License
GNU General Public License v3.0 or later

See [LICENSE](LICENSE) to see the full text.

CHANGELOG.md
### Version 0.2.1 (2020-06-06)
- Added configurations for simulations used in the articles

### Version 0.2 (2020-06-03)
- Added air travel option with a more realistic model. Travelers may now get infected during air travel.

covid19_sim/analysis.py
from enum import Enum
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import colors
import math
from textwrap import wrap
import json
import glob
import os

def monte_carlo_permutation_test(group_1, group_2, n_random):
    """
    An implementation of 2-tailed monte carlo permutation test.
    Compares the absolute value of the difference in means of 2 groups
    and returns a p-value.

    group_1: list for group 1
    group_2: list for group 2
    n_random: number of random samples generated from both groups
    """
    group_1_size, counter = len(group_1), 0
    observed_difference = np.abs(np.mean(group_1) - np.mean(group_2))
    pooled_group = np.concatenate([group_1, group_2])

    for i in range(n_random):
        np.random.shuffle(pooled_group)
        if np.abs(np.mean(pooled_group[:group_1_size]) - np.mean(pooled_group[group_1_size:])) >= observed_difference:
            counter += 1

    return counter / n_random

class Metric(Enum):
    TOTAL_INFECTED = 1
    MAX_INFECTED = 2
    TOTAL_UNNECESSARILY_QUARANTINED = 3
    DAYS_LOST_TO_UNNECESSARY_QUARANTINE = 4
    EPIDEMIC_DURATION = 5
    TOTAL_TEST_SESSIONS = 6

def create_custom_bar_chart(results, **kwargs):
    """Create bar chart for the experiments."""
    metrics = kwargs.get('metrics', [Metric.TOTAL_INFECTED, Metric.TOTAL_UNNECESSARILY_QUARANTINED])
    custom_labels = kwargs.get('custom_labels', [])

    #labels = results['labels']
    labels = ['Simulation ' + str(i) for i in range(1, len(results['labels']) + 1)]
    number_of_experiments = len(labels)
    if custom_labels and len(custom_labels) == number_of_experiments:
        labels = custom_labels
        
    width = math.ceil( 120 / len(labels) )
    if width < 20:
        width = 20
    labels = [ '\n'.join(wrap(l, width)) for l in labels ]

    data = []
    for metric in metrics:
        if metric == Metric.TOTAL_INFECTED:
            label = 'Total infected'
            values = results[Metric.TOTAL_INFECTED]
        if metric == Metric.MAX_INFECTED:
            label = 'Max infected'
            values = results[Metric.MAX_INFECTED]
        if metric == Metric.TOTAL_UNNECESSARILY_QUARANTINED:
            label = 'Total unnecessarily quarantined'
            values = results[Metric.TOTAL_UNNECESSARILY_QUARANTINED]
        if metric == Metric.EPIDEMIC_DURATION:
            label = 'Epidemic duration'
            values = results[Metric.EPIDEMIC_DURATION]
        data.append({
            'label': label,
            'values': values
        })
    
    x = np.arange(number_of_experiments)  # label locations
    width = 0.2 # the width of the bars

    fig, ax = plt.subplots()
    fig.set_size_inches(15, 5.5)
    metric_bars = []
    number_of_metrics = len(data)
    for i in range(number_of_metrics):
        metric_bars.append(ax.bar(x + (2*i - number_of_metrics + 1)*(width/2), data[i]['values'], width, label=data[i]['label']))

    # Add some text for labels, title and custom x-axis tick labels, etc.
    ax.set_ylabel('% of population', fontsize=18)
    ax.tick_params(axis='y', labelsize=14)
    ax.set_title('Comparison of different strategies', fontsize = 18)
    ax.set_xticks(x)
    ax.set_xticklabels(labels, fontsize=18)
    ax.legend(prop={'size': 18})

    def autolabel(rects):
        """Attach a text label above each bar in *rects*, displaying its height."""
        for rect in rects:
            height = rect.get_height()
            ax.annotate('{}'.format(height),
                        xy=(rect.get_x() + rect.get_width() / 2, height),
                        xytext=(0, 3),  # 3 points vertical offset
                        textcoords="offset points",
                        size = 18,
                        ha='center', va='bottom')
    for bars in metric_bars:
        autolabel(bars)
    
    fig.tight_layout()
    plt.xlim((-0.4, number_of_experiments - 0.3))
    plt.ylim((0, 109))
    plt.show()

def run_analysis(file_list):
    """Open files, load data and analyze them"""
    if not file_list:
        list_of_files = glob.glob('results/*.json')
        latest_file = max(list_of_files, key=os.path.getctime)
        if not latest_file:
            raise Exception('No files given to analyze!')
        file_list = [latest_file]
    
    results = []
    data = []
    for filename in file_list:
        file = open(filename, 'r')
        file_data = json.loads(file.read()) 
        file.close()
        data.append(file_data)
    
    # convert data in files into one single list
    n_elements = len(data[0])   # number of elements in each file_data
    for i in range(n_elements):
        merged = []
        for file_data in data:
            merged += file_data[i]
        results.append(merged)

    # ANALYSIS
    distributions = results[3]
    number_of_experiments = len(distributions)
    sim_properties = results[4]

    # check if population sizes of simulation data match
    width, height = sim_properties[0]['width'], sim_properties[0]['height']
    for i in range(number_of_experiments):
        if sim_properties[i]['width'] != width or sim_properties[i]['height'] != height:
            print('WARNING! POPULATION SIZES OF SIMULATIONS DO NOT MATCH!')

    # todo... hardcoded right now, change this
    total_infected_list = [[] for i in range(number_of_experiments)]
    epidemic_duration_list = [[] for i in range(number_of_experiments)]
    max_infected_list = [[] for i in range(number_of_experiments)]
    for j in range(number_of_experiments):
        for i in range(len(distributions[j])):
            total_infected_list[j].append(distributions[j][i][0])
            epidemic_duration_list[j].append(distributions[j][i][-2])
            max_infected_list[j].append(distributions[j][i][-1])

    avg_total_infected_list, avg_epidemic_duration_list, avg_max_infected_list = [], [], []
    for i in range(number_of_experiments):
        population_size = sim_properties[i]['width'] * sim_properties[i]['height']
        avg_total_infected_list.append(round(100*np.mean(total_infected_list[i]), 2))
        avg_epidemic_duration_list.append(round(np.mean(epidemic_duration_list[i]), 2))
        avg_max_infected_list.append(round(100*np.mean(max_infected_list[i])/population_size, 2))

    # compare results of experiments with each other
    for i in range(number_of_experiments):
        for j in range(i + 1, number_of_experiments):
            print()
            print('experiment ' + str(i + 1) + ' vs ' + str(j + 1))

            print('test for total infected: ' + str(avg_total_infected_list[i]) + ' vs ' + str(avg_total_infected_list[j]))
            p = monte_carlo_permutation_test(total_infected_list[i], total_infected_list[j], 5000)
            print('p = ' + str(p))

            print('test for epidemic duration: ' + str(avg_epidemic_duration_list[i]) + ' vs ' + str(avg_epidemic_duration_list[j]))
            p = monte_carlo_permutation_test(epidemic_duration_list[i], epidemic_duration_list[j], 5000)
            print('p = ' + str(p))

            print('test for max infected: ' + str(avg_max_infected_list[i]) + ' vs ' + str(avg_max_infected_list[j]))
            p = monte_carlo_permutation_test(max_infected_list[i], max_infected_list[j], 5000)
            print('p = ' + str(p))
    
    results_dict = {}
    results_dict['labels'] = results[0]
    results_dict[Metric.TOTAL_INFECTED] = results[1]
    results_dict[Metric.TOTAL_UNNECESSARILY_QUARANTINED] = results[2]
    results_dict[Metric.MAX_INFECTED] = avg_max_infected_list
    results_dict[Metric.EPIDEMIC_DURATION] = avg_epidemic_duration_list

    labels = []
    for sim in sim_properties:
        if not 'label' in sim:
            labels = []
            break
        labels.append(sim['label'])

    create_custom_bar_chart(results_dict, 
        metrics = [Metric.TOTAL_INFECTED, Metric.MAX_INFECTED], 
        custom_labels = labels
        )
    #create_custom_bar_chart(results_dict, metrics = [Metric.TOTAL_INFECTED, Metric.MAX_INFECTED, Metric.TOTAL_UNNECESSARILY_QUARANTINED, Metric.EPIDEMIC_DURATION])
    #create_custom_bar_chart(results_dict)